File: java/2 - Data Structures/Disjoint Set.java
Estructura de datos para modelar una colección de conjuntos disyuntos.
Permite determinar de manera eficiente a que conjunto pertenece un elemento,
si dos elementos se encuentran en un mismo conjunto y unir dos conjuntos disyuntos en un uno.

static int MAX = 100005;
    
    class Dsu{
        
        int dsu[] = new int[MAX];
        int size[] = new int[MAX];
        int numsets;
        
        
        Dsu(int n){
            for(int i = 0; i <= n; i++){
                dsu[i] = i;
                size[i] = 1;
            }
            numsets = n;
        }
        
        int find(int u){
            return (dsu[u] == u) ? u : (dsu[u] = find(dsu[u]));
        }
        
        void union(int a,int b){
            a = find(a);
            b = find(b);
            if(a == b)return;
            if(size[a] > size[b]){
                int c = a;
                a = b;
                b = c;
            }
            dsu[a] = b;
            size[b] += size[a];
            numsets--;
        }
        
        int sizeOf(int i) {
            return size[find(i)]; 
        }
        
    }

File: c++/2 - Data Structures/Disjoint Set.cpp
Estructura de datos para modelar una coleccion de conjuntos disyuntos.
Permite determinar de manera eficiente a que conjunto pertenece un elemento, si dos elementos se encuentran en un mismo conjunto y unir dos conjuntos en un uno.

struct dsu {
    vector<int> par, sz;
    int size; // Cantidad de conjuntos

    dsu(int n) : par(n), sz(n, 1) {
        size = n;
        iota(par.begin(), par.end(), 0);
    }
    // Busca el nodo representativo del conjunto de u
    int find(int u) {
        return par[u] == u ? u : (par[u] = find(par[u]));
    }
    // Une los conjuntos de u y v
    void unite(int u, int v) {
        u = find(u), v = find(v);
        if (u == v) return;
        if (sz[u] > sz[v]) swap(u, v);
        par[u] = v;
        sz[v] += sz[u];
        size--;
    }
    // Retorna la cantidad de elementos del conjunto de u
    int count(int u) { return sz[find(u)]; }
};
